xdsl/printer.py

```diff
@@ -349,7 +348,27 @@ def print_paramattr_parameters(
         self.print(">")
 
     def print_string_literal(self, string: str):
-        self.print(json.dumps(string))
+        # Common case, this makes a huge difference because we print one time instead
+        # of once per character.
+        # This makes a huge difference on programs with a lot of generic syntax.
+        if all(
+            (0x20 <= ord(char) <= 0x7E) and char not in ('"', "\\") for char in string
+        ):
+            self.print(f'"{string}"')
+            return
+        hexdigits = "0123456789ABCDEF"
+        self.print('"')
+        for char in string:
+            o = ord(char)
+            if char == "\\":
+                self.print("\\\\")
+            elif 0x20 <= o <= 0x7E and char != '"':
+                self.print(char)
+            else:
+                self.print("\\")
+                self.print(hexdigits[o >> 4])
+                self.print(hexdigits[o & 0x0F])
```

Comment:
Could we `f"{o:x}"` instead? This code here fails for larger unicode codepoints

Reply:
I disagree on not handling larger codepoints, MLIR does in a way
{o:x} prints the value in hexadecimal, but not the (MLIR) way we need to